logstore/lstoremem/addr_book.go

```diff
@@ -173,40 +340,35 @@ func (mab *memoryAddrBook) AddAddrs(t thread.ID, p peer.ID, addrs []ma.Multiaddr
 		return nil
 	}
 
-	s := mab.segments.get(p)
-	s.Lock()
-	defer s.Unlock()
-
-	amap, _ := s.getAddrs(t, p)
-	if amap == nil {
-		if s.addrs[t] == nil {
-			s.addrs[t] = make(map[peer.ID]map[string]*expiringAddr, 1)
-		}
-		amap = make(map[string]*expiringAddr, len(addrs))
-		s.addrs[t][p] = amap
-	}
-	exp := time.Now().Add(ttl)
-	for _, a := range addrs {
-		if a == nil {
-			log.Warnf("was passed nil multiaddr for %s", p)
-			continue
-		}
-		asBytes := a.Bytes()
-		x, found := amap[string(asBytes)] // won't allocate.
-		if !found {
-			// not found, save and announce it.
-			amap[string(asBytes)] = &expiringAddr{Addr: a, Expires: exp, TTL: ttl}
-			mab.subManager.BroadcastAddr(p, a)
-		} else {
-			// Update expiration/TTL independently.
-			// We never want to reduce either.
-			if ttl > x.TTL {
-				x.TTL = ttl
+	var addrChanged bool
+	mab.segments.processAddrs(t, p, true, true, func(amap map[string]*expiringAddr) {
+		exp := time.Now().Add(ttl)
+		for _, a := range addrs {
+			if a == nil {
+				log.Warnf("was passed nil multiaddr for %s", p)
+				continue
 			}
-			if exp.After(x.Expires) {
-				x.Expires = exp
+			asBytes := a.Bytes()
+			x, found := amap[string(asBytes)] // won't allocate.
+			if !found {
+				// not found, save and announce it.
+				amap[string(asBytes)] = &expiringAddr{Addr: a, Expires: exp, TTL: ttl}
+				mab.subManager.BroadcastAddr(p, a)
+				addrChanged = true
+			} else {
+				// Update expiration/TTL independently.
+				// We never want to reduce either.
+				if ttl > x.TTL {
+					x.TTL = ttl
+				}
+				if exp.After(x.Expires) {
+					x.Expires = exp
+				}
 			}
 		}
+	})
+	if addrChanged {
+		go mab.segments.updateEdge(t)
```

Comment:
similar to `HeadsEdges` approach - edges in the in-memory implementation are recomputed on the thread's info change as its pretty cheap operation here